### PR TITLE
add some concepts to blob compare

### PR DESCRIPTION
**Pull request by James-Mart**
Just for fun/practice.

Also increases readability, probably improves error messages...

### CODE REVIEW

File: libraries/psibase/common/include/psibase/blob.hpp [hunk 1]
```diff
@@ -4,16 +4,29 @@
 #include <algorithm>
 #include <type_traits>
 
+template <typename T>
+concept IsByteType =
+    std::is_same_v<std::decay_t<T>, char> || std::is_same_v<std::decay_t<T>, unsigned char>;
```
Comment (swatanabe) [inline libraries/psibase/common/include/psibase/blob.hpp:9]: `std::decay_t` should be applied outside this concept. I don't think that `IsByteType<char&>` should be true. Also, `remove_cvref` is usually preferable to `decay`.

File: libraries/psibase/common/include/psibase/blob.hpp [hunk 2]
```diff
@@ -4,16 +4,29 @@
...
+
+template <typename T>
+concept IsContainerLike = requires(T t) {
```
Comment (swatanabe) [inline libraries/psibase/common/include/psibase/blob.hpp:12]: Should use `const&`.

Comment (swatanabe) [inline libraries/psibase/common/include/psibase/blob.hpp:12]: This name is too generic for what it actually checks. Actually, if we're going to use concepts, it should be `std::ranges::contiguous_range<T> && std::ranges::sized_range<T>` (and use `std::ranges::data`/`std::ranges::size`).

File: libraries/psibase/common/include/psibase/blob.hpp (hunk 1 above)
Comment (James-Mart) [reply libraries/psibase/common/include/psibase/blob.hpp]: done/done

File: libraries/psibase/common/include/psibase/blob.hpp (hunk 2 above)
Comment (James-Mart) [reply libraries/psibase/common/include/psibase/blob.hpp]: no longer relevant

Comment (James-Mart) [reply libraries/psibase/common/include/psibase/blob.hpp]: This should now be addressed, what do you think?

**Review by swatanabe — APPROVED** [review]

**Pull request merged**